### PR TITLE
[app] add example on animating a view annotation along a routeline

### PR DESCRIPTION
**Pull request by tobrun**
To support @evil159 with a view annotation animation request on iOS, 
I created a small example to animate a view annotation across a route line geometry. 

ezgif com-gif-maker

While you do see some higher main thread render performance in the video, this is mostly coming from screenrecroding, below a screenshot without video recording, the animation also feels more smooth because of this.

image

### CODE REVIEW

File: app/src/main/res/layout/activity_view_annotation_animation.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <com.mapbox.maps.MapView
```
Comment (kiryldz) [inline app/src/main/res/layout/activity_view_annotation_animation.xml:6]: As we have just one element in the layout and it's `MapView` - perhaps it's easier to create in code making the activity code fully isolated and re-usable. Then this xml file could be totally removed.

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 2]
```diff
@@ -0,0 +1,1783 @@
...
+        // center camera around SF airport
+        setCamera(
+          CameraOptions.Builder()
+            .center(Point.fromLngLat(-122.3915, 37.6177))
+            .zoom(11.0)
+            .build()
+        )
+
+        // create a dummy view to animate
+        button = Button(binding.mapView.context)
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:76]: could re-use `item_callout_view.xml` from other examples and even e.g. update text dynamically when moving with new lat / lng.

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 3]
```diff
@@ -0,0 +1,1783 @@
...
+          ),
+        startValue.latitude() +
+          (endValue.latitude() - startValue.latitude()) * fraction
+      )
+    }
+
+  private companion object {
+    const val LINE_ID = "line"
+    const val SOURCE_ID = "source"
+    const val ROUTE_GEOJSON = """
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:142]: perhaps move to assets as it's quite long...

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 4]
```diff
@@ -0,0 +1,1783 @@
...
+    currentAnimator?.cancel()
+  }
+
+  private fun animateNextStep() {
+    if (count == routeCoordinateList.size - 1) {
+      count = 0
+    }
+    val currentPoint = routeCoordinateList[count]
+    val nextPoint = routeCoordinateList[count + 1]
+    val animateDuration = TurfMeasurement.distance(currentPoint, nextPoint) / totalLength * 30000
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:103]: 30000 to const and give meaningful name?

Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:103]: :+1: first part should be renamed to progress and 30000 should be renamed to total animation length

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+ * Example how to animate a view annotations on top of a route.
+ */
+class ViewAnnotationAnimationActivity : AppCompatActivity() {
+
+  private lateinit var viewAnnotationManager: ViewAnnotationManager
+  private lateinit var annotationView: View
+  private lateinit var textView: TextView
+  private var routeCoordinateList = mutableListOf<Point>()
+  private var totalLength = 0.0
+  private var count = 0
```
Comment (alexander-kulikovskii) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:41]:
```suggestion
  private var routeCoordinateIndex = 0
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+  override fun onStop() {
+    super.onStop()
+    currentAnimator?.cancel()
+  }
+
+  private fun animateNextStep() {
+    if (count == routeCoordinateList.size - 1) {
+      count = 0
+    }
```
Comment (alexander-kulikovskii) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:112]:
```suggestion
      routeCoordinateIndex %= (routeCoordinateList.size - 1)
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    val currentPoint = routeCoordinateList[count]
+    val nextPoint = routeCoordinateList[count + 1]
```
[the remark this answers is not in the document]
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:114]: that should be covered by resetting the routeCoordinateIndex in two lines before

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt (hunk 6 above)
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:112]: this is neat!

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt (hunk 5 above)
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:41]: :+1:

**Review by yunikkk — APPROVED** [review]

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    }
+  }
+
+  override fun onStop() {
+    super.onStop()
+    currentAnimator?.cancel()
+  }
+
+  private fun animateNextStep() {
+    currentRouteCoordinate %= (routeCoordinateList.size - 1)
```
Comment (alexander-kulikovskii) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:110]: for me `currentRouteCoordinate` should store coordinate as an object or pair.
so, maybe rename it.

```suggestion
    currentRouteCoordinateIndex %= (routeCoordinateList.size - 1)
```

**Review by alexander-kulikovskii — APPROVED** [review]

File: app/src/main/assets/sf_airport_route.geojson [hunk 9]
```diff
@@ -0,0 +1,1638 @@
+      {
+  "type": "FeatureCollection",
+  "features": [
```
Comment (pengdev) [inline app/src/main/assets/sf_airport_route.geojson:3]: nit: can we download this geojson from network, if it's taken from another js example?

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+          +lineLayer(LINE_ID, SOURCE_ID) {
+            lineColor(Color.parseColor(COLOR_PINK_HEX))
+            lineWidth(4.0)
+          }
+        }
+      ) {
+        // center camera around SF airport
+        setCamera(
+          CameraOptions.Builder()
+            .center(Point.fromLngLat(-122.3915, 37.6177))
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationAnimationActivity.kt:79]: nit: move to constant, one idea, maybe we should have some global util constants to host some frequently-used geo locations?

**Review by pengdev — APPROVED** [review]
LGTM % nits

**Pull request merged**